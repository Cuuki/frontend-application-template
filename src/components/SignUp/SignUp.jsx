import React, {useCallback, useState} from 'react';
import PropTypes from 'prop-types';
import {SignUpAddressGroupStyled} from './SignUp.styles';
import Form from '../UI/Form';
import FormField from '../UI/FormField';
import FormLabel from '../UI/FormLabel';
import TextInput from '../UI/TextInput';
import Button from '../UI/Button';
import CheckBox from '../UI/CheckBox';
import TextArea from '../UI/TextArea';

const NAME_MIN_LENGTH = 2;
const ADDRESS_MIN_LENGTH = 4;
const ZIP_LENGTH = 5;
const ZIP_PATTERN = '\\d{5}';
const EMAIL_PATTERN = '\\S+@\\S+\\.\\S+';
const PASSWORD_MIN_LENGTH = 6;
const PASSWORD_PATTERN = '^(?=.*[a-z])(?=(?:.*[0-9]){2}).*';

const SignUp = ({fieldValues, addField, clearFields, setFieldValidity}) => {
  const [isAddressHidden, setAddressVisibility] = useState(false);
  const addFieldCallback = useCallback(
    (name, customValidity = true) => event => {
      const {target} = event;

      addField({
        name,
        value: target.value,
        isValid: target?.validity?.valid && customValidity,
      });
    },
    [addField],
  );
  const getErrors = useCallback(
    () =>
      Object.keys(fieldValues).filter(
        fieldKey => !fieldValues[fieldKey]?.isValid,
      ),
    [fieldValues],
  );

  console.log(fieldValues.firstname);

  return (
    <Form
      handleSubmit={event => {
        event.preventDefault();
        clearFields();
        alert('Your data was sent to the server.');
      }}
    >
      <FormField
        labelComponent={<FormLabel describes="firstname">Firstname</FormLabel>}
      >
        <TextInput
          id="firstname"
          minLength={NAME_MIN_LENGTH}
          isRequired={true}
          initialValue={fieldValues.firstname?.value || ''}
          isValid={fieldValues.firstname?.isValid}
          handleChange={addFieldCallback('firstname')}
        />
      </FormField>

      <FormField
        labelComponent={<FormLabel describes="lastname">Lastname</FormLabel>}
      >
        <TextInput
          id="lastname"
          minLength={NAME_MIN_LENGTH}
          isRequired={true}
          initialValue={fieldValues.lastname?.value || ''}
          isValid={fieldValues.lastname?.isValid}
          handleChange={addFieldCallback('lastname')}
        />
      </FormField>

      <FormField
        labelComponent={<FormLabel describes="nickname">Nickname</FormLabel>}
      >
        <TextInput
          id="nickname"
          initialValue={fieldValues.nickname?.value || ''}
          isValid={fieldValues.nickname?.isValid}
          handleChange={addFieldCallback('nickname')}
        />
      </FormField>

      <FormField
        labelComponent={<FormLabel describes="email">Email</FormLabel>}
      >
        <TextInput
          id="email"
          type="email"
          isRequired={true}
          pattern={EMAIL_PATTERN}
          initialValue={fieldValues.email?.value || ''}
          isValid={fieldValues.email?.isValid}
          handleChange={addFieldCallback('email')}
        />
      </FormField>

      <FormField
        labelComponent={<FormLabel describes="password">Password</FormLabel>}
      >
        <TextInput
          id="password"
          type="password"
          minLength={PASSWORD_MIN_LENGTH}
          pattern={PASSWORD_PATTERN}
          isRequired={true}
          initialValue={fieldValues.password?.value || ''}
          isValid={fieldValues.password?.isValid}
          handleChange={event => {
            const {target} = event;
            const hasPasswordMatch =
              target.value === fieldValues.repeatPassword?.value;

            setFieldValidity({
              name: 'repeatPassword',
              isValid: hasPasswordMatch && target?.validity?.valid,
            });
            addFieldCallback('password', hasPasswordMatch)(event);
          }}
        />
      </FormField>

      <FormField
        labelComponent={
          <FormLabel describes="repeatPassword">Repeat password</FormLabel>
        }
      >
        <TextInput
          id="repeatPassword"
          type="password"
          minLength={PASSWORD_MIN_LENGTH}
          pattern={PASSWORD_PATTERN}
          isRequired={true}
          initialValue={fieldValues.repeatPassword?.value || ''}
          isValid={fieldValues.repeatPassword?.isValid}
          handleChange={event => {
            const {target} = event;
            const hasPasswordMatch =
              target.value === fieldValues.password?.value;

            setFieldValidity({
              name: 'password',
              isValid: hasPasswordMatch && target?.validity?.valid,
            });
            addFieldCallback('repeatPassword', hasPasswordMatch)(event);
          }}
        />
      </FormField>

      <CheckBox
        id="checkboxField"
        label="Show address"
        isChecked={!isAddressHidden}
        handleChange={() =>
          setAddressVisibility(previousVisibility => !previousVisibility)
        }
      />

      <SignUpAddressGroupStyled hidden={isAddressHidden}>
        <FormField
          labelComponent={<FormLabel describes="street">Street</FormLabel>}
        >
          <TextInput
            id="street"
            minLength={ADDRESS_MIN_LENGTH}
            isRequired={true}
            initialValue={fieldValues.street?.value || ''}
            isValid={fieldValues.street?.isValid}
            handleChange={addFieldCallback('street')}
          />
        </FormField>

        <FormField
          labelComponent={
            <FormLabel describes="apartment">Apartment</FormLabel>
          }
        >
          <TextInput
            id="apartment"
            isRequired={true}
            initialValue={fieldValues.apartment?.value || ''}
            isValid={fieldValues.apartment?.isValid}
            handleChange={addFieldCallback('apartment')}
          />
        </FormField>

        <FormField labelComponent={<FormLabel describes="zip">ZIP</FormLabel>}>
          <TextInput
            id="zip"
            minLength={ZIP_LENGTH}
            maxLength={ZIP_LENGTH}
            pattern={ZIP_PATTERN}
            isRequired={true}
            type="text"
            initialValue={fieldValues.zip?.value || ''}
            isValid={fieldValues.zip?.isValid}
            handleChange={addFieldCallback('zip')}
          />
        </FormField>

        <FormField
          labelComponent={<FormLabel describes="city">City</FormLabel>}
        >
          <TextInput
            id="city"
            isRequired={true}
            minLength={ADDRESS_MIN_LENGTH}
            initialValue={fieldValues.city?.value || ''}
            isValid={fieldValues.city?.isValid}
            handleChange={addFieldCallback('city')}
          />
        </FormField>
      </SignUpAddressGroupStyled>

      <FormField
        labelComponent={
          <FormLabel describes="additionalInformation">
            Additional information
          </FormLabel>
        }
      >
        <TextArea
          id="additionalInformation"
          initialValue={fieldValues.additionalInformation?.value || ''}
          isValid={fieldValues.additionalInformation?.isValid}
          handleChange={addFieldCallback('additionalInformation')}
        />
      </FormField>

      <Button type="submit" isDisabled={getErrors().length > 0}>
        Sign up now!
      </Button>
    </Form>
  );
};

SignUp.propTypes = {
  fieldValues: PropTypes.objectOf(
    PropTypes.exact({
      value: PropTypes.string,
      isValid: PropTypes.bool,
    }),
  ),
  addField: PropTypes.func.isRequired,
  clearFields: PropTypes.func.isRequired,
  setFieldValidity: PropTypes.func.isRequired,
};

SignUp.defaultProps = {
  fieldValues: {},
};

export default SignUp;
