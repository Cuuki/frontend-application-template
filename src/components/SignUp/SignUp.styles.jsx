import styled from 'styled-components';

export const SignUpAddressGroupStyled = styled.fieldset`
  display: ${({hidden}) => (hidden ? 'none' : 'block')};
`;
