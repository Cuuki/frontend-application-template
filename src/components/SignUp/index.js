import {connect} from 'react-redux';
import {createStructuredSelector} from 'reselect';
import {
  addField,
  removeField,
  setFieldValidity,
} from '../../redux/signUp/reducer/signUp.reducer';
import {signUpSelector} from '../../redux/signUp/selectors';
import SignUp from './SignUp';

const mapDispatchToProps = dispatch => ({
  addField: field => dispatch(addField(field)),
  removeField: fieldName => dispatch(removeField(fieldName)),
  setFieldValidity: field => dispatch(setFieldValidity(field)),
});

const mapStateToProps = createStructuredSelector({
  fieldValues: signUpSelector,
});

export default connect(mapStateToProps, mapDispatchToProps)(SignUp);
