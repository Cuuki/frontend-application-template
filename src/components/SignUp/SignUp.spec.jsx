import React from 'react';
import {render} from '@testing-library/react';
import SignUp from './SignUp';

describe('SignUp', () => {
  it('renders with no values set', () => {
    const {container} = render(
      <SignUp
        fieldValues={{}}
        addField={() => {}}
        removeField={() => {}}
        setFieldValidity={() => {}}
      />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders with values for firstname lastname and email', () => {
    const {container} = render(
      <SignUp
        fieldValues={{
          firstname: {
            value: 'John',
            isValid: true,
          },
          lastname: {
            value: 'Doe',
            isValid: true,
          },
          email: {
            value: 'test@example',
            isValid: false,
          },
        }}
        addField={() => {}}
        removeField={() => {}}
        setFieldValidity={() => {}}
      />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
