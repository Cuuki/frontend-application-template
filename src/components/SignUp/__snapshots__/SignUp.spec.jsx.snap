// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp renders with no values set 1`] = `
<form
  class="sc-gsTCUz"
>
  <div
    class="sc-dlfnbm"
  >
    <label
      class="sc-eCssSg"
      for="firstname"
    >
      Firstname
    </label>
    <input
      aria-invalid="true"
      class="sc-jSgupP"
      id="firstname"
      minlength="2"
      required=""
      type="text"
      value=""
    />
  </div>
  <div
    class="sc-dlfnbm"
  >
    <label
      class="sc-eCssSg"
      for="lastname"
    >
      Lastname
    </label>
    <input
      aria-invalid="true"
      class="sc-jSgupP"
      id="lastname"
      minlength="2"
      required=""
      type="text"
      value=""
    />
  </div>
  <div
    class="sc-dlfnbm"
  >
    <label
      class="sc-eCssSg"
      for="nickname"
    >
      Nickname
    </label>
    <input
      aria-invalid="true"
      class="sc-jSgupP"
      id="nickname"
      type="text"
      value=""
    />
  </div>
  <div
    class="sc-dlfnbm"
  >
    <label
      class="sc-eCssSg"
      for="email"
    >
      Email
    </label>
    <input
      aria-invalid="true"
      class="sc-jSgupP"
      id="email"
      pattern="\\\\S+@\\\\S+\\\\.\\\\S+"
      required=""
      type="email"
      value=""
    />
  </div>
  <div
    class="sc-dlfnbm"
  >
    <label
      class="sc-eCssSg"
      for="password"
    >
      Password
    </label>
    <input
      aria-invalid="true"
      class="sc-jSgupP"
      id="password"
      minlength="6"
      pattern="^(?=.*[a-z])(?=(?:.*[0-9]){2}).*"
      required=""
      type="password"
      value=""
    />
  </div>
  <div
    class="sc-dlfnbm"
  >
    <label
      class="sc-eCssSg"
      for="repeatPassword"
    >
      Repeat password
    </label>
    <input
      aria-invalid="true"
      class="sc-jSgupP"
      id="repeatPassword"
      minlength="6"
      pattern="^(?=.*[a-z])(?=(?:.*[0-9]){2}).*"
      required=""
      type="password"
      value=""
    />
  </div>
  <div
    class="sc-iBPRYJ"
  >
    <input
      checked=""
      class="sc-fubCfw"
      id="checkboxField"
      type="checkbox"
    />
    <label
      class="sc-eCssSg"
      for="checkboxField"
    >
      Show address
    </label>
  </div>
  <fieldset
    class="sc-bdfBwQ gjynrh"
  >
    <div
      class="sc-dlfnbm"
    >
      <label
        class="sc-eCssSg"
        for="street"
      >
        Street
      </label>
      <input
        aria-invalid="true"
        class="sc-jSgupP"
        id="street"
        minlength="4"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="sc-dlfnbm"
    >
      <label
        class="sc-eCssSg"
        for="apartment"
      >
        Apartment
      </label>
      <input
        aria-invalid="true"
        class="sc-jSgupP"
        id="apartment"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="sc-dlfnbm"
    >
      <label
        class="sc-eCssSg"
        for="zip"
      >
        ZIP
      </label>
      <input
        aria-invalid="true"
        class="sc-jSgupP"
        id="zip"
        maxlength="5"
        minlength="5"
        pattern="\\\\d{5}"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="sc-dlfnbm"
    >
      <label
        class="sc-eCssSg"
        for="city"
      >
        City
      </label>
      <input
        aria-invalid="true"
        class="sc-jSgupP"
        id="city"
        minlength="4"
        required=""
        type="text"
        value=""
      />
    </div>
  </fieldset>
  <div
    class="sc-dlfnbm"
  >
    <label
      class="sc-eCssSg"
      for="additionalInformation"
    >
      Additional information
    </label>
    <textarea
      aria-invalid="true"
      class="sc-pFZIQ"
      id="additionalInformation"
    />
  </div>
  <button
    class="sc-gKsewC gwmOxg"
    type="submit"
  >
    Sign up now!
  </button>
</form>
`;

exports[`SignUp renders with values for firstname lastname and email 1`] = `
<form
  class="sc-gsTCUz"
>
  <div
    class="sc-dlfnbm"
  >
    <label
      class="sc-eCssSg"
      for="firstname"
    >
      Firstname
    </label>
    <input
      aria-invalid="false"
      class="sc-jSgupP"
      id="firstname"
      minlength="2"
      required=""
      type="text"
      value="John"
    />
  </div>
  <div
    class="sc-dlfnbm"
  >
    <label
      class="sc-eCssSg"
      for="lastname"
    >
      Lastname
    </label>
    <input
      aria-invalid="false"
      class="sc-jSgupP"
      id="lastname"
      minlength="2"
      required=""
      type="text"
      value="Doe"
    />
  </div>
  <div
    class="sc-dlfnbm"
  >
    <label
      class="sc-eCssSg"
      for="nickname"
    >
      Nickname
    </label>
    <input
      aria-invalid="true"
      class="sc-jSgupP"
      id="nickname"
      type="text"
      value=""
    />
  </div>
  <div
    class="sc-dlfnbm"
  >
    <label
      class="sc-eCssSg"
      for="email"
    >
      Email
    </label>
    <input
      aria-invalid="true"
      class="sc-jSgupP"
      id="email"
      pattern="\\\\S+@\\\\S+\\\\.\\\\S+"
      required=""
      type="email"
      value="test@example"
    />
  </div>
  <div
    class="sc-dlfnbm"
  >
    <label
      class="sc-eCssSg"
      for="password"
    >
      Password
    </label>
    <input
      aria-invalid="true"
      class="sc-jSgupP"
      id="password"
      minlength="6"
      pattern="^(?=.*[a-z])(?=(?:.*[0-9]){2}).*"
      required=""
      type="password"
      value=""
    />
  </div>
  <div
    class="sc-dlfnbm"
  >
    <label
      class="sc-eCssSg"
      for="repeatPassword"
    >
      Repeat password
    </label>
    <input
      aria-invalid="true"
      class="sc-jSgupP"
      id="repeatPassword"
      minlength="6"
      pattern="^(?=.*[a-z])(?=(?:.*[0-9]){2}).*"
      required=""
      type="password"
      value=""
    />
  </div>
  <div
    class="sc-iBPRYJ"
  >
    <input
      checked=""
      class="sc-fubCfw"
      id="checkboxField"
      type="checkbox"
    />
    <label
      class="sc-eCssSg"
      for="checkboxField"
    >
      Show address
    </label>
  </div>
  <fieldset
    class="sc-bdfBwQ gjynrh"
  >
    <div
      class="sc-dlfnbm"
    >
      <label
        class="sc-eCssSg"
        for="street"
      >
        Street
      </label>
      <input
        aria-invalid="true"
        class="sc-jSgupP"
        id="street"
        minlength="4"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="sc-dlfnbm"
    >
      <label
        class="sc-eCssSg"
        for="apartment"
      >
        Apartment
      </label>
      <input
        aria-invalid="true"
        class="sc-jSgupP"
        id="apartment"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="sc-dlfnbm"
    >
      <label
        class="sc-eCssSg"
        for="zip"
      >
        ZIP
      </label>
      <input
        aria-invalid="true"
        class="sc-jSgupP"
        id="zip"
        maxlength="5"
        minlength="5"
        pattern="\\\\d{5}"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="sc-dlfnbm"
    >
      <label
        class="sc-eCssSg"
        for="city"
      >
        City
      </label>
      <input
        aria-invalid="true"
        class="sc-jSgupP"
        id="city"
        minlength="4"
        required=""
        type="text"
        value=""
      />
    </div>
  </fieldset>
  <div
    class="sc-dlfnbm"
  >
    <label
      class="sc-eCssSg"
      for="additionalInformation"
    >
      Additional information
    </label>
    <textarea
      aria-invalid="true"
      class="sc-pFZIQ"
      id="additionalInformation"
    />
  </div>
  <button
    class="sc-gKsewC jkrbqk"
    disabled=""
    type="submit"
  >
    Sign up now!
  </button>
</form>
`;
